File: app/src/main/java/com/plbear/iweight/view/DataAdapter.java
package com.plbear.iweight.view;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.SparseArray;

import com.plbear.iweight.Data.Data;
import com.plbear.iweight.Data.DataManager;
import com.plbear.iweight.Utils.ILog;
import com.plbear.iweight.Utils.OldUtils;
import com.plbear.iweight.Utils.SPUtils;
import com.plbear.iweight.model.settings.SettingsActivity;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

/**
 * Created by yanyongjun on 16/11/5.
 */

public class DataAdapter {
    private String TAG = "DataAdapter:";
    private ArrayList<Data> mDataList = new ArrayList<Data>();
    private ArrayList<Data> mShowList = new ArrayList<Data>();
    private float mWeightBiggest = 0;
    private float mWeightSmallest = 0;
    private long mBeginTime = 0;
    private long mEndTime = 0;

    private ArrayList<DataChangeListener> mListener = new ArrayList<DataChangeListener>();
    private Context mContext = null;
    //目标体重值
    private float mTagetWeight = -1;
    private DataManager mDataManager = null;
    private final int DEFAULT_NUMS = 7;
    private int mShowPointCount = DEFAULT_NUMS;

    private int mShowBeginId;

    private boolean mShowAllData = false;

    public DataAdapter(Context context) {
        init(context);
    }


    public int getShowPointCount() {
        return mShowAllData ? mShowList.size() : mShowPointCount;
    }

    public void setShowAllData(boolean flag) {
        mShowAllData = flag;
        notifyDataSetChange();
    }

    public void setTag(String tag) {
        TAG += tag;
        ILog.e(TAG, "setTag");
    }

    /**
     * init 初始化各个参数值，在构造方法中进行调用
     *
     * @returnd
     */
    private void init(Context context) {
        mContext = context;
        SharedPreferences sp = SPUtils.Companion.getSP(mContext);
        mTagetWeight = sp.getFloat(SettingsActivity.PREFERENCE_KEY_SET_TARGET_WEIGHT, -1) *
                OldUtils.getValueUnit(context);

        DataManager db = DataManager.getInstance(context);
        ArrayList<Data> list = db.queryAll();
        mDataList.clear();
        mDataList.addAll(list);

        setShowList(mDataList);

        mDataManager = DataManager.getInstance(mContext);
    }

    public void setShowNum(int num) {
        mShowPointCount = num;
        notifyDataSetChange();
    }

    public interface DataChangeListener {
        public void onChange();
    }


    public void registerDataListener(DataChangeListener listener) {
        mListener.add(listener);
    }


    /**
     * 通知注册Adapter Lisntenr的类进行刷新
     */
    private void notifyDataChange() {
        SharedPreferences sp = SPUtils.Companion.getSP(mContext);
        mTagetWeight = sp.getFloat(SettingsActivity.PREFERENCE_KEY_SET_TARGET_WEIGHT, -1)
                * OldUtils.getValueUnit(mContext);
        for (DataChangeListener listener : mListener) {
            listener.onChange();
        }
    }

    public void notifyDataSetChange() {
        ILog.e(TAG, "notifyDataSetChange");
        DataManager db = DataManager.getInstance(mContext);
        ArrayList<Data> list = db.queryAll();
        mDataList.clear();
        mDataList.addAll(list);

        setShowList(list);

        notifyDataChange();
    }

    public void notifyDataPosSetChange(float moveLength) {
        //TODO
    }

    private void setShowList(ArrayList<Data> allList) {
        if (allList == null || allList.size() == 0) {
            mShowList.clear();
            mWeightBiggest = 0;
            mWeightSmallest = 0;
            mBeginTime = 0;
            mEndTime = 0;
            return;
        }
        ILog.e(TAG, "setShowList:" + mShowPointCount + ":mShowAllData:" + mShowAllData);
        if (mShowAllData) {
            mShowList = allList;
            mShowBeginId = mDataList.get(0).getId();
            SparseArray<String> list = resolveShowList();
            mEndTime = Long.valueOf(list.get(1));
            mBeginTime = Long.valueOf(list.get(2));
            mWeightBiggest = Float.valueOf(list.get(3));
            mWeightSmallest = Float.valueOf(list.get(4));
            return;
        }
        if (mDataList.size() <= mShowPointCount) {
            mShowBeginId = mDataList.get(0).getId();
        } else {
            mShowBeginId = mDataList.get(mDataList.size() - (int) mShowPointCount).getId();
        }

        mShowList.clear();
        boolean add = false;
        int count = 0;
        for (Data data : allList) {
            if (data.getId() == mShowBeginId) {
                add = true;
            }
            if (add) {
                count++;
                if (count > mShowPointCount) {
                    break;
                }
                mShowList.add(data);
            }
        }
        SparseArray<String> list = resolveShowList();
        mEndTime = Long.valueOf(list.get(1));
        mBeginTime = Long.valueOf(list.get(2));
        mWeightBiggest = Float.valueOf(list.get(3));
        mWeightSmallest = Float.valueOf(list.get(4));
    }

    private void sort() {
        Collections.sort(mDataList, new Comparator<Data>() {
            @Override
            public int compare(Data data, Data data2) {
                if (data.getTime() > data2.getTime()) {
                    return 1;
                } else {
                    return -1;
                }
            }
        });
    }

    /**
     * 得到目标体重，这个体重刷新后的数据来源应用该是唯一的，就是这里
     *
     * @return
     */
    public float getTargetWeight() {
        return mTagetWeight;
    }

    public ArrayList<Data> getShowDataList() {
        return mShowList;
    }

    /**
     * 返回要展示的list的第一个id
     *
     * @return
     */
    public int getShowDataStartId() {
        //TODO
        return mShowBeginId;
    }

    public long getTimeSmallest() {
        return mBeginTime;
    }

    public long getTimeBiggest() {
        return mEndTime;
    }

    public float getWeightBiggest() {

        if (mTagetWeight == -1) {
            return getTrueWeightBiggest();
        }
        return (mWeightBiggest > mTagetWeight ? mWeightBiggest : mTagetWeight) + 5;
    }

    public float getWeightSmallest() {
        if (mTagetWeight == -1) {
            return getTrueWeightSmallest();
        }
        return (mWeightSmallest < mTagetWeight ? mWeightSmallest : mTagetWeight) - 5;
    }

    public float getTrueWeightBiggest() {
        return mWeightBiggest + 5;
    }

    public float getTrueWeightSmallest() {
        return mWeightSmallest - 5;
    }


    public SparseArray<String> resolveShowList() {
        SparseArray<String> list = new SparseArray<String>();
        long timeBiggest = 0;
        long timeSmallest = Long.MAX_VALUE;
        float weightBiggest = 0;
        float weightSmallest = Float.MAX_VALUE;
        for (Data data : mShowList) {
            if (timeBiggest <= data.getTime()) {
                timeBiggest = data.getTime();
            }
            if (timeSmallest >= data.getTime()) {
                timeSmallest = data.getTime();
            }
            if (weightBiggest <= data.getWeight()) {
                weightBiggest = data.getWeight();
            }
            if (weightSmallest >= data.getWeight()) {
                weightSmallest = data.getWeight();
            }
        }
        list.put(1, String.valueOf(timeBiggest));
        list.put(2, String.valueOf(timeSmallest));
        list.put(3, String.valueOf(weightBiggest));
        list.put(4, String.valueOf(weightSmallest));
        return list;
    }


    public float getHeight() {
        return getWeightBiggest() - getWeightSmallest();
    }


}

File: app/src/main/java/com/plbear/iweight/view/DataAdapter.kt
package com.plbear.iweight.view

import android.content.Context
import android.content.SharedPreferences
import android.util.SparseArray
import com.plbear.iweight.data.Data

import com.plbear.iweight.data.DataManager
import com.plbear.iweight.utils.MyLog
import com.plbear.iweight.utils.SPUtils
import com.plbear.iweight.model.settings.SettingsActivity
import com.plbear.iweight.utils.Utils

import java.util.ArrayList
import java.util.Collections
import java.util.Comparator

/**
 * Created by yanyongjun on 16/11/5.
 */

class DataAdapter(context: Context) {
    private var TAG = "DataAdapter:"
    private val mDataList = ArrayList<Data>()
    var showDataList = ArrayList<Data>()
        private set
    private var mWeightBiggest = 0f
    private var mWeightSmallest = 0f
    var timeSmallest: Long = 0
        private set
    var timeBiggest: Long = 0
        private set

    private val mListener = ArrayList<DataChangeListener>()
    private var mContext: Context? = null
    //目标体重值
    /**
     * 得到目标体重，这个体重刷新后的数据来源应用该是唯一的，就是这里
     *
     * @return
     */
    var targetWeight = -1f
        private set
    private var mDataManager: DataManager? = null
    private val DEFAULT_NUMS = 7
    private var mShowPointCount = DEFAULT_NUMS

    /**
     * 返回要展示的list的第一个id
     *
     * @return
     */
    //TODO
    var showDataStartId: Int = 0
        private set

    private var mShowAllData = false


    val showPointCount: Int
        get() = if (mShowAllData) showDataList.size else mShowPointCount

    val weightBiggest: Float
        get() = if (targetWeight == -1f) {
            trueWeightBiggest
        } else (if (mWeightBiggest > targetWeight) mWeightBiggest else targetWeight) + 5

    val weightSmallest: Float
        get() {
            return if (targetWeight == -1f) {
                trueWeightSmallest
            } else (if (mWeightSmallest < targetWeight) mWeightSmallest else targetWeight) - 5
        }

    val trueWeightBiggest: Float
        get() = mWeightBiggest + 5

    val trueWeightSmallest: Float
        get() = mWeightSmallest - 5


    val height: Float
        get() = weightBiggest - weightSmallest

    init {
        init(context)
    }

    fun setShowAllData(flag: Boolean) {
        mShowAllData = flag
        notifyDataSetChange()
    }

    fun setTag(tag: String) {
        TAG += tag
        MyLog.e(TAG, "setTag")
    }

    /**
     * init 初始化各个参数值，在构造方法中进行调用
     *
     * @returnd
     */
    private fun init(context: Context) {
        mContext = context
        val sp = SPUtils.getSp()
        targetWeight = sp.getFloat(SettingsActivity.PREFERENCE_KEY_SET_TARGET_WEIGHT, -1f) * Utils.getValueUnit()

        val db = DataManager.getInstance(context)
        val list = db!!.queryAll()
        mDataList.clear()
        mDataList.addAll(list)

        setShowList(mDataList)

        mDataManager = DataManager.getInstance(mContext)
    }

    fun setShowNum(num: Int) {
        mShowPointCount = num
        notifyDataSetChange()
    }

    interface DataChangeListener {
        fun onChange()
    }


    fun registerDataListener(listener: DataChangeListener) {
        mListener.add(listener)
    }


    /**
     * 通知注册Adapter Lisntenr的类进行刷新
     */
    private fun notifyDataChange() {
        val sp = SPUtils.getSp()
        targetWeight = sp.getFloat(SettingsActivity.PREFERENCE_KEY_SET_TARGET_WEIGHT, -1f) * Utils.getValueUnit()
        for (listener in mListener) {
            listener.onChange()
        }
    }

    fun notifyDataSetChange() {
        MyLog.e(TAG, "notifyDataSetChange")
        val db = DataManager.getInstance(mContext)
        val list = db!!.queryAll()
        mDataList.clear()
        mDataList.addAll(list)

        setShowList(list)

        notifyDataChange()
    }

    fun notifyDataPosSetChange(moveLength: Float) {
        //TODO
    }

    private fun setShowList(allList: ArrayList<Data>?) {
        if (allList == null || allList.size == 0) {
            showDataList.clear()
            mWeightBiggest = 0f
            mWeightSmallest = 0f
            timeSmallest = 0
            timeBiggest = 0
            return
        }
        MyLog.e(TAG, "setShowList:$mShowPointCount:mShowAllData:$mShowAllData")
        if (mShowAllData) {
            showDataList = allList
            showDataStartId = mDataList[0].id
            val list = resolveShowList()
            timeBiggest = java.lang.Long.valueOf(list.get(1))!!
            timeSmallest = java.lang.Long.valueOf(list.get(2))!!
            mWeightBiggest = java.lang.Float.valueOf(list.get(3))!!
            mWeightSmallest = java.lang.Float.valueOf(list.get(4))!!
            return
        }
        if (mDataList.size <= mShowPointCount) {
            showDataStartId = mDataList[0].id
        } else {
            showDataStartId = mDataList[mDataList.size - mShowPointCount.toInt()].id
        }

        showDataList.clear()
        var add = false
        var count = 0
        for (data in allList) {
            if (data.id === showDataStartId) {
                add = true
            }
            if (add) {
                count++
                if (count > mShowPointCount) {
                    break
                }
                showDataList.add(data)
            }
        }
        val list = resolveShowList()
        timeBiggest = java.lang.Long.valueOf(list.get(1))!!
        timeSmallest = java.lang.Long.valueOf(list.get(2))!!
        mWeightBiggest = java.lang.Float.valueOf(list.get(3))!!
        mWeightSmallest = java.lang.Float.valueOf(list.get(4))!!
    }

    private fun sort() {
        Collections.sort<Data>(mDataList) { data, data2 ->
            if (data.time > data2.time) {
                1
            } else {
                -1
            }
        }
    }


    fun resolveShowList(): SparseArray<String> {
        val list = SparseArray<String>()
        var timeBiggest: Long = 0
        var timeSmallest = java.lang.Long.MAX_VALUE
        var weightBiggest = 0f
        var weightSmallest = java.lang.Float.MAX_VALUE
        for (data in showDataList) {
            if (timeBiggest <= data.time) {
                timeBiggest = data.time
            }
            if (timeSmallest >= data.time) {
                timeSmallest = data.time
            }
            if (weightBiggest <= data.weight) {
                weightBiggest = data.weight
            }
            if (weightSmallest >= data.weight) {
                weightSmallest = data.weight
            }
        }
        list.put(1, timeBiggest.toString())
        list.put(2, timeSmallest.toString())
        list.put(3, weightBiggest.toString())
        list.put(4, weightSmallest.toString())
        return list
    }


}
